File: VB/MvpvmNavigation/Presenter/MainFormPresenter.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.XtraEditors
Imports MvpvmNavigation.ViewModels
Imports DevExpress.Mvvm
Imports MvpvmNavigation.Model
Imports System.Windows.Forms
Imports DevExpress.XtraBars.Ribbon

Namespace MvpvmNavigation.Presenter
    Public Class MainFormPresenter
        Implements IDisposable

        Private modulesPanel As PanelControl
        Private viewModelCore As MainFormViewModel
        Public Sub New(ByVal modulesPanel As PanelControl, ByVal viewModel As MainFormViewModel)
            Me.modulesPanel = modulesPanel
            Me.viewModelCore = viewModel
            AddHandler Me.ViewModel.SelectedModuleTypeChanged, AddressOf ViewModel_SelectedModuleTypeChanged
        End Sub
        Public Sub Dispose() Implements IDisposable.Dispose
            RemoveHandler ViewModel.SelectedModuleTypeChanged, AddressOf ViewModel_SelectedModuleTypeChanged
        End Sub
        Public ReadOnly Property ViewModel() As MainFormViewModel
            Get
                Return viewModelCore
            End Get
        End Property
        Private Sub ViewModel_SelectedModuleTypeChanged(ByVal sender As Object, ByVal e As System.EventArgs)
            Dim mainForm As RibbonForm = TryCast(modulesPanel.FindForm(), RibbonForm)
            If mainForm IsNot Nothing Then
                mainForm.Ribbon.UnMergeRibbon()
                If mainForm.Ribbon.StatusBar IsNot Nothing Then
                    mainForm.Ribbon.StatusBar.UnMergeStatusBar()
                End If
            End If
            modulesPanel.Controls.Clear()
            Dim [module] As Control = TryCast(ViewModel.GetModule(ViewModel.SelectedModuleType), Control)
            ViewModelHelper.EnsureModuleViewModel([module], ViewModel)
            If [module] IsNot Nothing Then
                [module].Dock = DockStyle.Fill
                modulesPanel.Controls.Add([module])
            End If
            If mainForm IsNot Nothing Then
                Dim ribbonModule As IRibbonModule = TryCast([module], IRibbonModule)
                If ribbonModule IsNot Nothing Then
                    mainForm.Ribbon.MergeRibbon(ribbonModule.Ribbon)
                    If mainForm.Ribbon.StatusBar IsNot Nothing Then
                        mainForm.Ribbon.StatusBar.MergeStatusBar(ribbonModule.Ribbon.StatusBar)
                    End If
                End If
            End If
        End Sub
    End Class
End Namespace

File: CS/MvpvmNavigation/Presenter/MainFormPresenter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.XtraEditors;
using MvpvmNavigation.ViewModels;
using DevExpress.Mvvm;
using MvpvmNavigation.Model;
using System.Windows.Forms;
using DevExpress.XtraBars.Ribbon;

namespace MvpvmNavigation.Presenter {
    public class MainFormPresenter : IDisposable {
        PanelControl modulesPanel;
        MainFormViewModel viewModelCore;
        public MainFormPresenter(PanelControl modulesPanel, MainFormViewModel viewModel) {
            this.modulesPanel = modulesPanel;
            this.viewModelCore = viewModel;
            ViewModel.SelectedModuleTypeChanged += ViewModel_SelectedModuleTypeChanged;
        }
        public void Dispose() {
            ViewModel.SelectedModuleTypeChanged -= ViewModel_SelectedModuleTypeChanged;
        }
        public MainFormViewModel ViewModel {
            get { return viewModelCore; }
        }
        void ViewModel_SelectedModuleTypeChanged(object sender, System.EventArgs e) {
            RibbonForm mainForm = modulesPanel.FindForm() as RibbonForm;
            if (mainForm != null) {
                mainForm.Ribbon.UnMergeRibbon();
                if (mainForm.Ribbon.StatusBar != null)
                    mainForm.Ribbon.StatusBar.UnMergeStatusBar();
            }
            modulesPanel.Controls.Clear();
            Control module = ViewModel.GetModule(ViewModel.SelectedModuleType) as Control;
            ViewModelHelper.EnsureModuleViewModel(module, ViewModel);
            if (module != null) {
                module.Dock = DockStyle.Fill;
                modulesPanel.Controls.Add(module);
            }
            if (mainForm != null) {
                IRibbonModule ribbonModule = module as IRibbonModule;
                if (ribbonModule != null) {
                    mainForm.Ribbon.MergeRibbon(ribbonModule.Ribbon);
                    if (mainForm.Ribbon.StatusBar != null)
                        mainForm.Ribbon.StatusBar.MergeStatusBar(ribbonModule.Ribbon.StatusBar);
                }
            }
        }
    }
}
